### PR TITLE
Selenium System Tests #224

### LINKED ISSUE #224: [Testing] Integration (system) tests
As a Developer, I would like to have system tests for all functionalities of the website including Search, file upload, signup, login, dataset upload, displaying, opening saved graph and saving graph, profile page.

**Description:**
Currently search, Graph are unit tested but they are lacking integration test. The integration test will test all components as a whole (system)

**Note**

**Tasks Involved:**

- [x]  Develop integration test for Search, file upload, Graph, 
- [x] Develop all UI tests for all user stories since sprint 4

**Acceptance Criteria**
- All tests must ensure the functionality of the developed parts of the system (Search, file upload, graph)

### LINKED ISSUE #318: AT-1 Graph Feature Acceptance Test with Changes
##### Here there are all tests needed for this feature

Graph epic consists of sub-stories that were mostly implemented

Tests | Variable Inputs | Expected Result | Status
-- | -- | -- | --
Access the home page | Launch the website on localhost | Homepage is shown |  
Access the graph page | Navigate to the graph page from the side menu | The graph page appears |  
Find dataset to add to graph | Click “ADD DATASET TO GRAPH” button, search for some values, like “year = 1990” and click “SEARCH DATABASE” | A search form will appear, after searching for values a list of Datasets will be displayed on the form |  
Add datasets to graph | Select one or more datasets from the list shown and click  “ADD TO GRAPH” | Datasets will be added to graph |  
Change the display settings | Click “Settings” button, choose “X Variable” and “Y variable” to be compared and shown on graph | Datasets will be shown on the graph according to the X and Y variables selected |  
Hide/Show dataset on graph | Click on the “eye” icon on a specific dataset to hide it, clicking again will show the dataset again | A specific dataset will be hidden or shown |  
Remove a dataset from the graph | Click on the “Trash bin” icon to remove a dataset from the graph | A specific dataset will be removed from the graph |  
Change the scaling of the graph, and return it to default view | Use the scales on the top and right sides to change the X and Y scales on the graph respectively, clicking on the “-” blue circle in the top right corner of the graph will return the scales to default | The scales of the graph will change |

### LINKED ISSUE #319: AT-3 Sign Up/Log In/Log Out
This acceptance test tests the features to sign up, log in, and log out

Tests | Variable Inputs | Expected Result | Status
-- | -- | -- | --
Access the home page | Launch the website on localhost | Homepage is shown |  
Navigate to Log in page | Click “LOG IN” button in the top right corner to navigate to the login page | Login page is shown |  
Sign up for a new account | Click “Don’t have an account? Sign Up” link to navigate to sign up page | The Sign up page form will be displayed |  
Fill the information and sign up | Fill out all the required fields and then click “SIGN UP” button | The account will get created and the user will get rerouted to Log in with the new account |  
Log in with new account | Fill in the email address and password of the newly created account and click “LOG IN” | The user will be logged in |  
Sign Out | Click “SIGN OUT” button on the top right corner to sign out of your account | The user will get signed out |

### LINKED ISSUE #320: AT-4 Account Information
Tests | Variable Inputs | Expected Result | Status
-- | -- | -- | --
Access the home page | Launch the website on localhost | Homepage is shown |  
Navigate to Log in page | Click “LOG IN” button in the top right corner to navigate to the login page | Login page is shown |  
Log in to your account | Enter the email address and password in the required fields and when done click “LOG IN” | Redirected to Home page and the user is signed in |  
Navigate to Profile page | Navigate to the profile page from the side menu | The profile page appears |  
Edit your profile | Click “EDIT PROFILE” button to edit some personal information | A list of editable information will be shown |  
Save your new settings | Click “SAVE” after editing or “CANCEL” to cancel your changes | The settings will be saved and updated or cancelled based on the choice |  
Sign Out | Click “SIGN OUT” button on the top right corner to sign out of your account | The user will get signed out |  

### Demo Steps

### LINKED ISSUE #321: AT-5 Saving a Graph
Tests | Variable Inputs | Expected Result | Status
-- | -- | -- | --
Access the home page | Launch the website on localhost | Homepage is shown |  
Navigate to Log in page | Click “LOG IN” button in the top right corner to navigate to the login page | Login page is shown |  
Log in to your account | Enter the email address and password in the required fields and when done click “LOG IN” | Redirected to Home page and the user is signed in |  
Access the graph page | Navigate to the graph page from the side menu | The graph page appears |  
Find dataset to add to graph | Click “ADD DATASET TO GRAPH” button, search for some values, like “year = 1990” and click “SEARCH DATABASE” | A search form will appear, after searching for values a list of Datasets will be displayed on the form |  
Add datasets to graph | Select one or more datasets from the list shown and click  “ADD TO GRAPH” | Datasets will be added to graph |  
Change the display settings | Click “Settings” button, choose “X Variable” and “Y variable” to be compared and shown on graph | Datasets will be shown on the graph according to the X and Y variables selected |  
Save the graph in its current state | After editing and modifying the graph fill the “Graph Name” field with a name of your choosing and click on “SAVE GRAPH” button | A message will be displayed confirming the success of saving the graph |  
Access the profile page | Navigate to the profile page from the side menu | The profile page appears |  
Access the view favourites tab | Navigate to the second tab called “VIEW FAVOURITES” and click on it | The saved Graphs page appears |  
Display the saved graph | Find your recently saved graph by the name and click on it | Reroute to graph page and the graph will be displayed in its saved condition |

### LINKED ISSUE #328: AT - 6 Password Reset
Tests | Variable Inputs | Expected Result | Status
-- | -- | -- | --
Access the home page | Launch the website on localhost | Homepage is shown |  
Navigate to Login page | Click “LOG IN” button in the top right corner to navigate to the login page | Login page is shown |  
Navigate to Forget Password form | Click “Forgot password?” link | The forget password form will pop up |  
Fill the form | Enter the email address associated with the account to recover password then click “FORGOT PASSWORD” button | A notification will confirm an email sent to reset password |  
Find the link to reset password | Navigate to your own email account, find the email from “noreply@databoom.com” , copy and paste the link sent in your browser | New password page is shown |  
Fill out the new password form | Fill out a new password and confirm it then click “RESET PASSWORD” button | A notification will confirm the password is changed |

### LINKED ISSUE #330: AT-2 File Upload with Changes
Tests | Variable Inputs | Expected Result | Status
-- | -- | -- | --
Access the home page | Launch the website on localhost | Homepage is shown |  
Navigate to File Upload page | Open the menu on the by clicking the top left button, then click “File Upload” | File Upload page is shown |  
Choose a file to upload | Click “Choose File” button and choose the file to be uploaded and click “Open” button when done | A file will be chosen and ready to be uploaded |  
Upload the file | Click “UPLOAD” button to upload the file | A notification will confirm the upload of the file and the page will be redirected to Upload Dataset page |  
Review/Edit information and save dataset | The files’s data is mapped on the page and can be edited or corrected, when done click “SAVE DATASET” button on the bottom | The dataset will be saved in the database |

### LINKED ISSUE #332: AT-7 Admin Review
Tests | Variable Inputs | Expected Result | Status
-- | -- | -- | --
Access the home page | Launch the website on localhost | Homepage is shown |  
Navigate to Login | Click “LOG IN” button in th top right corner to navigate to the login page | Login page is shown |  
Log in to your admin account | Enter the admin’s email address and password in the required fields and when finished clock “LOG IN” | Redirected to Home page and the admin is signed in |  
Navigate to Admin Review page | Choose a dataset from the list of unapproved datasets on the left side | The dataset information will be mapped on the page |  
Review/Edit dataset | Review the dataset information or click “EDIT” to be able to modify the data, when finished click “APPROVE DATASET” button in the bottom | A notification will appear confirming the dataset was accepted and it will be removed from the list of unapproved datasets |  
Flag a dataset | Enter a Flagged Comment in the required field and click “FLAG” button to flag the selected dataset | A notification will appear confirming the dataset was flagged |  
Delete a dataset | Click “DELETE” button | A notification will appear confirming the deletion of dataset and the dataset will be removed from the list of unapproved datasets |

### PR DESCRIPTION
**Pull request by laithAwad**
The file Selenium-System-Tests contains new ui tests that covers all the functions of the website

### CODE REVIEW

File: Server/upload/95a14afff5747afd941776100c754b0d [hunk 1]
```diff
@@ -0,0 +1,240 @@
+{
+  "reference": {
```
Comment (ds161833) [inline Server/upload/95a14afff5747afd941776100c754b0d:2]: remove the server files that were generated in the process but one that GD added, ill send you the id of it

File: Client/src/Components/Graph/GraphControls/SaveGraphStateForm.tsx [hunk 2]
```diff
@@ -51,7 +51,7 @@ export const SaveGraphStateForm = (props: IProps) => {
               <Field name="name" label='Graph Name' component={MuiTextFieldFormik} />
             </Grid>
             <Grid item sm={4}>
-              <Button type="submit" variant="contained" color="primary">Save Graph</Button>
+              <Button id="SaveGraph" type="submit" variant="contained" color="primary">Save Graph</Button>
```
Comment (ratulkai93) [inline Client/src/Components/Graph/GraphControls/SaveGraphStateForm.tsx:54]: nice work

File: Server/upload/95a14afff5747afd941776100c754b0d (hunk 1 above)
Comment (ds161833) [reply Server/upload/95a14afff5747afd941776100c754b0d:2]: all but this: 
image

File: Client/src/Components/Profile/UserDetailSection/PasswordSection/PasswordChangeForm.tsx [hunk 3]
```diff
@@ -44,7 +44,7 @@ export default function PasswordChangeForm(props: IProps) {
             </Grid>
           </Grid>
         </Box>
-        <Button variant="contained" color="primary" type="submit">Confirm Password</Button>
+        <Button id="ConfirmPasswordButton" variant="contained" color="primary" type="submit">Confirm Password</Button>
```
Comment (ratulkai93) [inline Client/src/Components/Profile/UserDetailSection/PasswordSection/PasswordChangeForm.tsx:47]: clean

File: Server/upload/2f5ac5899264abca839cda9d9ec4c0d9 [hunk 4]
```diff
@@ -0,0 +1,240 @@
+{
```
Comment (ratulkai93) [inline Server/upload/2f5ac5899264abca839cda9d9ec4c0d9:1]: remove this file please

File: Server/upload/95a14afff5747afd941776100c754b0d [hunk 5]
```diff
@@ -0,0 +1,240 @@
+{
```
Comment (ratulkai93) [inline Server/upload/95a14afff5747afd941776100c754b0d:1]: remove this one too

File: Server/src/routes/getDatasetRouter.ts [hunk 6]
```diff
@@ -0,0 +1,35 @@
+import { Request, Response, Router } from 'express';
```
Comment (ds161833) [inline Server/src/routes/getDatasetRouter.ts:1]: wait what is this file, how come it shows as added

File: Server/upload/c702a389adaaafa8beca2540ead07a22 [hunk 7]
```diff
@@ -0,0 +1,240 @@
+{
```
Comment (ratulkai93) [inline Server/upload/c702a389adaaafa8beca2540ead07a22:1]: this too

File: Server/upload/c9d1319df83bda59fcfd6081af2046b3 [hunk 8]
```diff
@@ -0,0 +1,240 @@
+{
```
Comment (ratulkai93) [inline Server/upload/c9d1319df83bda59fcfd6081af2046b3:1]: this too

File: Server/upload/e1c6fe433e5d3a2a439e1f9664225b77 [hunk 9]
```diff
@@ -0,0 +1,240 @@
+{
```
Comment (ratulkai93) [inline Server/upload/e1c6fe433e5d3a2a439e1f9664225b77:1]: this too LOL

**Review by ratulkai93 — APPROVED** [review]
looks good overall, great job
just make sure to remove the extra uploaded files ;)

**Review by ds161833 — APPROVED** [review]
nice, i see they are attached by ids, good job

**Review by silent-po2 — APPROVED** [review]
A+

File: Server/src/routes/getDatasetRouter.ts (hunk 6 above)
Comment (laithAwad) [reply Server/src/routes/getDatasetRouter.ts:1]: Removed

**Comment by laithAwad** [comment]
https://user-images.githubusercontent.com/26638534/106405760-4dee2980-6405-11eb-8200-fd3de38bc35c.mov


https://user-images.githubusercontent.com/26638534/106405787-55153780-6405-11eb-98b0-e0553553bd5b.mov


https://user-images.githubusercontent.com/26638534/106405797-5ba3af00-6405-11eb-9060-218784da42fe.mov


https://user-images.githubusercontent.com/26638534/106405801-5cd4dc00-6405-11eb-9d36-227ed8671d1d.mov


https://user-images.githubusercontent.com/26638534/106405808-61999000-6405-11eb-8fd4-7f8b46843e05.mov


https://user-images.githubusercontent.com/26638534/106405819-6b22f800-6405-11eb-9506-4f9d54982067.mov


https://user-images.githubusercontent.com/26638534/106405847-7b3ad780-6405-11eb-92ae-13fbe1b8db25.mov


https://user-images.githubusercontent.com/26638534/106405851-7ece5e80-6405-11eb-9400-9bbc0ef0a314.mov


https://user-images.githubusercontent.com/26638534/106405858-8261e580-6405-11eb-828d-d0fdefde2d37.mov


https://user-images.githubusercontent.com/26638534/106405863-8aba2080-6405-11eb-80f5-d4f2b9bd1a17.mov


https://user-images.githubusercontent.com/26638534/106405874-93aaf200-6405-11eb-84ec-419c981eca04.mov

Added videos showing all the test cases

**Pull request merged**